packages-exp/messaging-exp/src/index.sw.ts

```diff
@@ -15,16 +15,50 @@
  * limitations under the License.
  */
 
+import '@firebase/installations-exp';
+
+import { ERROR_FACTORY, ErrorCode } from './util/errors';
+import {
+  onNotificationClick,
+  onPush,
+  onSubChange
+} from './listeners/sw-listeners';
+
 import { FirebaseMessaging } from './interfaces/public-types';
+import { MessagingService } from './messaging-service';
+import { ServiceWorkerGlobalScope } from './util/sw-types';
+import { getApp } from '@firebase/app-exp';
+import { getMessaging } from './api';
+import { isSwSupported } from './api/isSupported';
 import { registerMessaging } from './helpers/register';
-import '@firebase/installations-exp';
 
 export { onBackgroundMessage, getMessaging } from './api';
+export { isSwSupported as isSupported } from './api/isSupported';
 
+declare const self: ServiceWorkerGlobalScope;
 declare module '@firebase/component' {
   interface NameServiceMapping {
     'messaging-exp': FirebaseMessaging;
   }
 }
 
+// Top-level 'await' requires 'module' option set to 'esnext' or 'system', and 'target' option
+// set to 'es2017' or higher. For compatibility, use async expression here.
+void (async () => {
+  if (!(await isSwSupported())) {
+    throw ERROR_FACTORY.create(ErrorCode.UNSUPPORTED_BROWSER);
+  }
+})();
+
 registerMessaging();
+
+const messaging = getMessaging(getApp());
+self.addEventListener('push', e => {
+  e.waitUntil(onPush(e, messaging as MessagingService));
+});
+self.addEventListener('pushsubscriptionchange', e => {
+  e.waitUntil(onSubChange(e, messaging as MessagingService));
+});
+self.addEventListener('notificationclick', e => {
+  e.waitUntil(onNotificationClick(e));
+});
```

Comment:
I don't think we want to auto-initialize messaging on user's behalf. It also doesn't work well if user creates multiple app/messaging instances. Can we move the listeners registration to the `getMessaging()` function, so they are registered when developer initializes messaging?

It also means we will have 2 versions of `getMessaging()`, one for the client and one of the sw.